src/canopy/llm/openai.py

```diff
@@ -206,3 +207,135 @@ async def agenerate_queries(self,
+                               messages: Messages,
+                               function: Function,
+                               *,
+                               max_tokens: Optional[int] = None,
+                               model_params: Optional[dict] = None,) -> dict:
+        """
+        This function enforces the model to respond with a specific function call.
+
+        To read more about this feature, see: https://platform.openai.com/docs/guides/gpt/function-calling
+
+        Note: this function is wrapped in a retry decorator to handle transient errors.
+
+        Args:
+            messages: Messages (chat history) to send to the model.
+            function: Function to call. See canopy.llm.models.Function for more details.
+            max_tokens: Maximum number of tokens to generate. Defaults to None (generates until stop sequence or until hitting max context size).
+            model_params: Model parameters to use for this request. Defaults to None (uses the default model parameters).
+                          Overrides the default model parameters if set on initialization.
+                          For example, you can pass: {"temperature": 0.9, "top_p": 1.0} to override the default temperature and top_p.
+                          see: https://platform.openai.com/docs/api-reference/chat/create
+
+        Returns:
+            dict: Function call arguments as a dictionary.
+
+        Usage:
+            >>> from canopy.llm import AzureOpenAILLM
+            >>> from canopy.llm.models import Function, FunctionParameters, FunctionArrayProperty
+            >>> from canopy.models.data_models import UserMessage
+            >>> llm = AzureOpenAILLM()
+            >>> messages = [UserMessage(content="I was wondering what is the capital of France?")]
+            >>> function = Function(
+            ...     name="query_knowledgebase",
+            ...     description="Query search engine for relevant information",
+            ...     parameters=FunctionParameters(
+            ...         required_properties=[
+            ...             FunctionArrayProperty(
+            ...                 name="queries",
+            ...                 items_type="string",
+            ...                 description='List of queries to send to the search engine.',
+            ...             ),
+            ...         ]
+            ...     )
+            ... )
+            >>> llm.enforced_function_call(messages, function)
+            {'queries': ['capital of France']}
+        """  # noqa: E501
+
+        model_params_dict: Dict[str, Any] = deepcopy(self.default_model_params)
+        model_params_dict.update(model_params or {})
+
+        function_dict = cast(ChatCompletionToolParam, function.dict())
+
+        chat_completion = self._client.chat.completions.create(
+            model=self.model_name,
+            messages=[m.dict() for m in messages],
+            functions=[function_dict],
```

Comment:
A. Are we sure that if you deploy your model with the latest GPT version (e.g. `gpt-3.5-turbo-1106`), the `tools=` argument is still not supported?  
If it does - we should probably remove the duplication and use `OpenAILLM`'s implementation directly.

Reply:
B. I noticed that the default `gpt-3.5-turbo` version on Azure OpenAI services is `0301`, which doesn't support function calling. It means that for many users - this code is going to error out, and OpenAI's error message is not very intuitive.  

We should probably add a `try: except openai.NotFoundError:` clause around this call, and re-raise with a clearer error message that tells the user that the model version defined in their deployment doesn't support function calling.